037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+Because we are talking about auto-rebalancing, the corresponding field could also be not limited to the rebalancing on scaling but a more general full continuous” auto-rebalancing to enable on the cluster.
+
+The `spec.kafka.autoRebalance` field allows to specify a list of auto-rebalancing modes with the corresponding templates to be used.
+To specify when running the auto-rebalancing, a `spec.kafka.autoRebalance.mode` field could be used to specify if running auto-rebalancing only on adding or removing brokers, or a full rebalance is requested.
+Having a full auto-rebalancing is out of scope for this proposal, the field is added for enabling a future implementation.
+
+The template to be used is referenced by a new mandatory `spec.kafka.autoRebalance.template` field and the cluster operator uses that template to create a corresponding `KafkaRebalance` custom resource.
+If the referenced `KafkaRebalanceTemplate` custom resource doesn't exist, the cluster operator will fail the reconcile.
+
+In the following example, a user can decide to have different rebalancing templates when adding or removing brokers and when they want a full rebalancing during the normal cluster execution.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      # using the same rebalancing template when adding or removing brokers
+      - mode: add-brokers
+        template: my-add-remove-brokers-rebalancing-template
+      - mode: remove-brokers
+        template: my-add-remove-brokers-rebalancing-template  
+      # using a different template for a full rebalancing  
+      - mode: full
+        template: my-full-rebalancing-template
+```
+
+The user can even decide to not have full auto-rebalancing and, for example, using different templates when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      # using a specific template when adding brokers
+      - mode: add-brokers
+        template: my-add-brokers-rebalancing-template
+      # using a different template when removing brokers  
+      - mode: remove-brokers
+        template: my-remove-brokers-rebalancing-template
+      # NOTE: no full auto-rebalancing   
+```
+
+The user could even decide to have auto-rebalancing only when adding or removing brokers but not for both.
+
+This way provides the greatest flexibility to describe in which case to run the auto-rebalancing and with which configuration.
+
+### Auto-rebalancing on scaling up
+
+Following the operator flow when auto-rebalancing on scaling up.
+
+1. The user scales up the cluster by increasing the `spec.kafka.replicas` field.
+2. The `KafkaReconciler` takes care of it by rolling the existing brokers first and then scaling up the cluster as usual, by adding additional brokers.
```

Comment:
The existing brokers should not be rolled anymore in most cases when StrimziPodSets are used. So maybe you can make it a bit more generic and avoid mentioning it explciitely.